prusti-viper/src/encoder/procedure_encoder.rs

```diff
@@ -4985,32 +5054,78 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
         self.encode_copy_value_assign(encoded_lhs, encoded_val, ty, location)
     }
 
-    fn encode_assign_array_len(
+    fn encode_assign_sequence_len(
         &mut self,
         encoded_lhs: vir::Expr,
         place: &mir::Place<'tcx>,
         dst_ty: ty::Ty<'tcx>,
         location: mir::Location,
     ) -> SpannedEncodingResult<Vec<vir::Stmt>> {
         trace!(
-            "[enter] encode_assign_array_len(place={:?}, ty={:?})",
+            "[enter] encode_assign_sequence_len(place={:?}, ty={:?})",
             place,
             dst_ty,
         );
         let span = self.mir_encoder.get_span_of_location(location);
+        let (encoded_place, mut stmts, place_ty, ..) = self.encode_place(place).with_span(span)?;
+        let ty_name = self.encoder.encode_type_predicate_use(place_ty)
+            .with_span(span)?;
+        match place_ty.kind() {
+            ty::TyKind::Array(_, ref ty_len) => {
+                // extract the length from the array type
+                let len = self.encoder.const_eval_intlike(&ty_len.val).unwrap()
+                    .to_u64().unwrap();
+
+                stmts.extend(
+                    self.encode_copy_value_assign(
+                        encoded_lhs,
+                        len.into(),
+                        dst_ty,
+                        location,
+                    )?
+                );
+            },
+            ty::TyKind::Slice(..) => {
```

Comment:
Is it possible to add a test case that executes this branch? I'm not sure how to generate a `mir::Rvalue::Len`. Maybe it's automatically generated by the compiler when it checks the bounds of slice indexes.

Reply:
added a testcase, indeed bounds checks are exercising this case.